src/lib/cimbar_js/cimbar_recv_js.cpp

```diff
@@ -0,0 +1,219 @@
+#include "encoder/escrow_buffer_writer.h"
+#include "extractor/Extractor.h"
+#include "fountain/fountain_decoder_sink.h"
+#include "serialize/str_join.h"
+#include "util/Timer.h"
+
+#include <opencv2/opencv.hpp>
+
+#include <algorithm>
+#include <memory>
+
+
+namespace {
+
+	std::shared_ptr<fountain_decoder_sink> _sink;
+	std::string _reporting;
+
+	cv::Mat _debugFrame;
+
+	TimeAccumulator _tScanExtract;
+	TimeAccumulator _tImgDecode;
+
+	// settings
+	unsigned _colorBits = 2;
+	int _modeVal = 68;
+
+	bool legacy_mode()
+	{
+		return _modeVal == 4;
+	}
+
+	cv::UMat get_rgb(void* imgdata, int width, int height, int type)
+	{
+		cv::UMat img;
+		switch (type)
+		{
+			case 12:
+			{
+				img = cv::Mat(height * 3/2, width, CV_8UC1, imgdata).getUMat(cv::ACCESS_RW).clone();
+				cv::cvtColor(img, img, cv::COLOR_YUV2RGB_NV12); // 12 or 21 :hmm:
+				return img;
+			}
+			case 420:
+			{
+				img = cv::Mat(height * 3/2, width, CV_8UC1, imgdata).getUMat(cv::ACCESS_RW).clone();
+				cv::cvtColor(img, img, cv::COLOR_YUV420p2RGB);
+				return img;
+			}
+			default:
+				break;
+		}
+
+		int cvtype = type==4? CV_8UC4 : CV_8UC3;
+		img = cv::Mat(height, width, cvtype, imgdata).getUMat(cv::ACCESS_RW).clone();
+		if (type == 4)
+			cv::cvtColor(img, img, cv::COLOR_RGBA2RGB);
+		return img;
+	}
+}
+
+extern "C" {
+
+unsigned cimbard_get_report(uchar* buff, unsigned maxlen)
+{
+	int len = std::min<unsigned>(_reporting.size(), maxlen);
+	if (len == 0)
+		return 0;
+	std::copy(_reporting.data(), _reporting.data()+len, buff);
+	return len;
+}
+
+unsigned cimbard_get_debug(uchar* buff, unsigned maxlen)
+{
+	int len = std::min<unsigned>(_debugFrame.dims*_debugFrame.cols*_debugFrame.rows, maxlen);
+	if (len == 0)
+		return 0;
+	std::copy(_debugFrame.data, _debugFrame.data+len, buff);
+	return len;
+}
+
+int cimbard_get_bufsize()
+{
+	unsigned chunksPerFrame = cimbar::Config::fountain_chunks_per_frame(cimbar::Config::symbol_bits() + _colorBits, legacy_mode());
+	unsigned chunkSize = cimbar::Config::fountain_chunk_size(
+				cimbar::Config::ecc_bytes(),
+				cimbar::Config::symbol_bits() + _colorBits,
+				legacy_mode()
+	);
+	return chunksPerFrame * chunkSize;
+}
+
+int cimbard_scan_extract_decode(uchar* imgdata, unsigned imgw, unsigned imgh, int format, uchar* bufspace, unsigned bufsize)
+{
+	if (format <= 0)
+		format = 3;
+	if (imgw == 0 or imgh == 0)
+		return -1;
+
+	unsigned chunksPerFrame = cimbar::Config::fountain_chunks_per_frame(cimbar::Config::symbol_bits() + _colorBits, legacy_mode());
+	unsigned chunkSize = cimbar::Config::fountain_chunk_size(
+				cimbar::Config::ecc_bytes(),
+				cimbar::Config::symbol_bits() + _colorBits,
+				legacy_mode()
+	);
+	// early bail if bufsize doesn't match config params (fountain chunk size * count)
+	if (bufsize < chunkSize * chunksPerFrame)
+		return -2;
+
+	// interface to take the aligned output buffers of chunkSize and dump them into bufspace
+	escrow_buffer_writer ebw(bufspace, chunksPerFrame, chunkSize);
```

Comment:
This feels slightly awkward, but the gist is that we know there's a hard upper bound on the size of an output from a call to `decode_fountain()` -- and we also know we're only making one call here. So we can have this object (which functions a bit like an fstream) hold the output buffer memory location, which allows the decoded bytes to land exactly where we need them.